### ISSUE TITLE
why should we use "softmax" twice in "run.py"? 

### ISSUE THREAD

**Issue by 9p15p** [seoungwugoh/RGMP#16]
1.  in 71 line: ` msv_E2[sc] = F.softmax(e2[0], dim=1)[:,1].data.cpu()`

2.  in 116 line: `all_E[:,:,f+1] = F.softmax(Variable(all_E[:,:,f+1]), dim=1).data`

Why should we use "softmax" twice?

**Comment by seoungwugoh** [comment]
Hi @9p15p,
For multi-object inference, we first estimate the probability map for each object. Then, we combine them with the soft aggregation.  
The first softmax (L71) is for the estimation of each object, and the second softmax (L116) is for implementing the soft aggregation operation.

**Comment by ryancll** [comment]
Hi Seoung,

I'm very interested in your soft aggregation approach for multi-object task and I can understand your mindset after reading your paper. I'm wondering how this approach improve the performance compared with winner-take-all approach? 

Thank you!

**Comment by seoungwugoh** [comment]
Hi @ryancll,

It is because maintaining uncertainty is important during propagation. After Soft aggregation the results is a probability map, while, with winner-take-all approach, it is a binary map. I think soft probability map gives the network more information to address challenges during propagation.
